### ISSUE TITLE
Release 0.2 release bricked TS100 (originally v3.42)

### ISSUE THREAD

**Issue by hugovincent** [Ralim/IronOS-dfu#11]
Hi, I seem to have bricked my TS100 using release 0.2. I'm not sure what's gone wrong since everything seemed to proceed as in the instructions. I first installed `runtime.hex` using stock mass storage bootloader. It successfully booted with the IronOS logo and DFU v0.2 shown on screen. I then backed up the stock bootloader:
```
% sudo dfu-util -d 28e9:0189 -U backup.bin  -s 0x08000000:0x4000

Password:
dfu-util 0.11

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2021 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

Opening DFU capable USB device...
Device ID 28e9:0189
Device DFU version 011a
Claiming USB DFU Interface...
Setting Alternate Interface #0 ...
Determining device status...
DFU state(2) = dfuIDLE, status(0) = No error condition is present
DFU mode device DFU version 011a
Device returned transfer size 1024
DfuSe interface name: "Internal Flash "
Upload	[=========================] 100%        16384 bytes
Upload done.

% md5 backup.bin 
MD5 (backup.bin) = cd1038179b3fda32e75d6f300d153cb3
```

That hash matches that listed for TS100 v3.42 in https://github.com/Ralim/IronOS-dfu/blob/mainline/docs/BackUp.md#checking-your-bootloader-backup-is-valid so it looks good. Then I power cycled, then installed the DFU bootloader:

```
% dfu-util -D bootloader.dfu
dfu-util 0.11

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2021 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

Match vendor ID from file: 0483
Match product ID from file: df11
Multiple alternate interfaces for DfuSe file
Opening DFU capable USB device...
Device ID 28e9:0189
Device DFU version 011a
Claiming USB DFU Interface...
Setting Alternate Interface #0 ...
Determining device status...
DFU state(2) = dfuIDLE, status(0) = No error condition is present
DFU mode device DFU version 011a
Device returned transfer size 1024
DfuSe interface name: "Internal Flash "
File contains 1 DFU images
Parsing DFU image 1
Target name: ST...
Image for alternate setting 0, (1 elements, total size = 5384)
Setting Alternate Interface #0 ...
Parsing element 1, address = 0x08000000, size = 5376
Erase   	[=========================] 100%         5376 bytes
Erase    done.
Download	[=========================] 100%         5376 bytes
Download done.
Done parsing DfuSe file
```
Again, seems good (no errors, `Download done.` seen). At this point I power cycled the iron by unplugging and replugging it in, and it seems to be bricked. Any ideas? (I have an ST-Link and will make up a cable for SWD reflashing... I'm reporting as an issue in case there is a bug or doc improvement that could be made).

Thanks!

**Comment by Ralim** [comment]
Hmmm I'm not certain how this failed.
If you have the st-link; can you please try and dump your flash before you write the new bootloader.
Would be amazing to get a diff of the two to see what failed to be written to the flash.

**Comment by hugovincent** [comment]
@Ralim I finally connected up the ST-Link, and found that read-out protection was enabled. I disabled it using OpenOCD, and presumably as a consequence of that, the image read back blank (all 0xff). So, sorry, no clues.

**Comment by discip** [comment]
@hugovincent
Did you manage to get your **`TS100`** to run again?

@Ralim
As you might recall I had a similar issue from the beginning. (exactly as @hugovincent stated above, except for the device being a **`TS80P`**)
And if I remember right, I had a problems flashing it even using the ST-Link until I somehow removed some protection, unfortunately I am not sure what exactly it was! 😓
But none the less maybe I will be lucky enough to get the information you need from an **`TS80`**, if that is enough.

btw:
> HW V1.2
> DFU 3.50
> APP 1.27
>
> ```
> Open On-Chip Debugger
> > stm32f1x options_read 0
> option byte register = 0x3fffffe
> write protection register = 0xffffffff
> read protection: on
> watchdog: software
> stop mode: no reset generated upon entry
> standby mode: no reset generated upon entry
> user data = 0xffff
> ```

_Originally posted by @jamesturton in https://github.com/Ralim/IronOS-dfu/issues/9#issuecomment-1163490992_

**Comment by discip** [comment]
@hugovincent
Did you manage to get your **`TS100`** to run again?

@Ralim
As you might recall I had a similar issue from the beginning. (exactly as @hugovincent stated above, except for the device being a **`TS80P`**)
And if I remember right, I had problems flashing it even using the ST-Link until I somehow removed some protection, unfortunately I am not sure what exactly it was! 😓
But none the less maybe I will be lucky enough to get the information you need from an **`TS80`**, if that is enough.

btw:
> HW V1.2
> DFU 3.50
> APP 1.27
>
> ```
> Open On-Chip Debugger
> > stm32f1x options_read 0
> option byte register = 0x3fffffe
> write protection register = 0xffffffff
> read protection: on
> watchdog: software
> stop mode: no reset generated upon entry
> standby mode: no reset generated upon entry
> user data = 0xffff
> ```

_Originally posted by @jamesturton in https://github.com/Ralim/IronOS-dfu/issues/9#issuecomment-1163490992_

**Comment by hugovincent** [comment]
> @hugovincent Did you manage to get your **`TS100`** to run again?

Yes I did, it was fine after reflashing over SWD. (Or, it would have been if I didn't accidentally reassemble the MCU module the wrong way around, burning out diode D1. After replacing that, it works well again).

**Comment by discip** [comment]
## Great! 😃👍
Then it would be appropriate to close this issue. 😊

Thanks in advance!

**Comment by discip** [comment]
@hugovincent
## Great! 😃👍
Then it would be appropriate to close this issue. 😊

Thanks in advance!

**Comment by hugovincent** [comment]
@discip you're welcome to close it if you want of course; I would rather leave it open as the cause has not been identified or fixed.

**Comment by Ralim** [comment]
The most likely thing here is that your device had the flash protection level set for the first part of the flash, which means any attempt to write to that area will trigger a device erase. A bit tough to work around really though. Could try re-writing all of this to run from ram but that is a fair bit more effort.

**Comment by hugovincent** [comment]
Instead of running from RAM, could you add a check for the flash protection and abort the installation if it's set? That way at least it would fail safe.

**Comment by Devnol** [comment]
I just bricked my TS100 the same way, thankfully the stm daughter board is a 2.51A which from what I understand connects swd to the usb port, so I can probably easily connect to that. Could you please give me some rough instructions on how you went to flash back the bootloader to get the iron working? I don't have an stlink unfortunately but I do have a pi pico I could probably use as an swd debugger, if that's what's needed. Thanks in advance.

**Comment by Devnol** [comment]
Looks like I managed to get openocd working and connected with gdb, what do I need to do now to flash the bootloader back?

**Comment by discip** [comment]
@Devnol
I used the ST-Link solution, but I am not sure what exactly I did do to get it running back then.

All I know is that you need to flash the previously backed up bootloader and then repeat what you did before:
Copy the **`runtime.hex`** file and then the **`bootloader.dfu`**.

Sorry, but that's really all I can remember. 😓

I recommend waiting until @Ralim comes back from vacation. 😊

**Comment by Devnol** [comment]
Yeah first of all I didn't take a backup of my bootloader because I saw this issue and that I should've done so after the fact 😓 Can't I just flash the ironos dfu directly? I got openocd and gdb up and running, I just don't know what commands I have to run. I guess I can wait for ralim to return lol, but thanks a lot anyway.

**Comment by discip** [comment]
Does your **`TS80P`** run on a ST-chip?
Or is it GD?

**Comment by Devnol** [comment]
I have a TS100, uses an stm32fsomething.
Also check the ironos repo, I think I fixed the translation (had to open a new pr)

**Comment by discip** [comment]
Ah, ok?
The only devices I have at hand are a **`TS80P`** and a **`Pinecil_V1`**.
So you will have to wait than.

**Comment by Devnol** [comment]
That's fine, I'm not in a hurry to solder anything right now, I just can't believe how I played myself like a fool yesterday by overwriting my bootloader...

**Comment by Devnol** [comment]
Hey, looks like I managed to get the ironos bootloader on here with a little help from the voidstarlab discord, for future reference, I had to run openocd with the iron attached to my probe (in this case a pi pico) then, inside the folder where the firmware is get an openocd shell with telnet 4444 and run the following:
```
init
reset halt
srm32f1x unlock 0
reset halt
flash write_image bootloader.hex
flash verify_image bootloader.hex
```
then optionally repeat the same for the running firmware too.
So now I'm in the ralim dfu, what do I need to run to flash a new ironos image or boot logo?

**Comment by discip** [comment]
@Devnol
To flash the firmware:
1. Download the latest [firmware](https://github.com/Ralim/IronOS/suites/11587617034/artifacts/600523881),
2. unzip the corresponding xxx.dfu
3. and run for example:
   ```
   dfu-util -D TS100_EN.dfu   
   ```
To flash a boot-logo:
1. Download the latest [release](https://github.com/Ralim/IronOS-Meta/releases/download/latest/miniware.zip),
2. unzip the desired xxx.dfu
3. and run for example:
   ```
   dfu-util -D IronOS.gif.dfu   
   ```

**Comment by Devnol** [comment]
perfect, thanks!

**Comment by Devnol** [comment]
You should probably make a note on the docs that flashing the ironos dfu bootloader without unlocking the mcu first will cause it to erase itself and be bricked

**Comment by discip** [comment]
Is your iron up and running again?

**Comment by Devnol** [comment]
yep, I just wanted to change my fw to greek because I flashed the en one with openocd. In any case, should someone else accidentally brick their iron these should be the steps to fix it after getting to an openocd shell.

**Comment by discip** [comment]
> You should probably make a note on the docs that flashing the ironos dfu bootloader without unlocking the mcu first will cause it to erase itself and be bricked

Sorry to have to let you down, but since I have no deeper understanding of that matter I am leaving this to @Ralim, instead of creating a write up about things I do not know.

I'm sorry to disappoint you but since I don't have a deeper understanding of the matter I'll leave that to @Ralim instead of writing a note about things I don't understand thoroughly enough.

**Comment by discip** [comment]
> You should probably make a note on the docs that flashing the ironos dfu bootloader without unlocking the mcu first will cause it to erase itself and be bricked

I'm sorry to disappoint you but since I don't have a deeper understanding of the matter I'll leave that to @Ralim instead of writing a note about things I don't understand thoroughly enough.

**Comment by Devnol** [comment]
Don't worry, I fixed mine lol. Just saying that you should probably warn people from now on to avoid this bricking again, you don't have to do it now and it doesn't need to be specifically you, just saying that ralim should make a note of that

**Comment by discip** [comment]
I got that.
Just needed to let him know. 😊

**Comment by Ralim** [comment]
There already is a warning in the docs that 3.45 onwards has this issue. Will update it to indicate some earlier DFU versions are also affected.

**Comment by tdrkDev** [comment]
+1 on this issue. I've bricked my TS100 with the same v3.42 DFU after flashing v0.2 bootloader. I've already ordered an ST-Link to repair it :(
The only good thing is that I've backed up stock bootloader according to the warning in docs.
I've not seen any warnings about this issue for old DFU firmwares in docs (https://github.com/Ralim/IronOS-dfu/blob/mainline/docs/Bootloader.md), @Ralim, maybe I've read them careless btw

UPD: I've successfully repaired it after soldering st-link wires to small CPU board and flashing IronOS DFU after unlocking flash.

**Comment by kenblu24** [comment]
Managed to brick it as well after a seemingly successful bootloader flash, so I had to do it with ST-Link.

Didn't see a pinout for those pads for hooking the daughterboard up to SWD, so here. 
![_DSC6824](https://github.com/Ralim/IronOS-dfu/assets/7890808/b18fb3f1-cd99-426d-9400-c0b1a5219703)
Highly recommend probing and doing a continuity check using the image on the right before soldering to those pads. I am obviously not responsible if you break your iron even more.

Pinout on the right taken from GD32F103Tx QFN36 pinout on datasheet
![_DSC6823-2](https://github.com/Ralim/IronOS-dfu/assets/7890808/711a5ea4-433a-42bf-b50d-77ec03da2ee3)

1. Soldered thin solid core wires to the pads and put some female dupont headers on the other ends
2. hooked it up to a crappy ST-Link knockoff to program the IronOS-dfu bootloader.hex file (STM32CubeProgrammer didn't work, so I used the deprecated ST-Link software)
3. desoldered the wires, put the daughterboard back on the iron, connected via the micro-usb and saw it appear in a `dfu-util -l`
4. flashed the `TS100_XX.dfu` file using dfu-util
5. ???
6. profit

**Comment by Devnol** [comment]
Just as a side note for anyone else reading this, the above process is necessary for daughter boards older than v2.51A, as that added swd to the usb port, d+ being swdio and d- being swclk. In that case all you'll need is a usb breakout and a microusb cable, so if you don't have a second soldering iron it's no big deal.

**Comment by AreYouLoco** [comment]
Damn mine is 2.47A using STM32 chip. So I need second soldering Iron to fix it. Should I flash .hex file or .bin file? Runtime first or with the external programmer I can do bootloader already? I have ST-Link v2 that saved my ass already once.

Fun fact! I had broken TS100 in some drawer. It didnt heat up. I kept it to fix it one day. But guess what I swapped daughter boards from the one that didnt heat up as the software there was ok to the bricked one. Ofc flashed IronOS straigjt away and e voila I have one working now. So now I may solder pins and unbrick the one that all began with.

So for now no messing with bootloaders until both are working.

**Comment by Ralim** [comment]
Hi, @AreYouLoco Can you provide more details sorry, did you flash the board without heeding the warning that on newer units dont let you do this or was this another issue?

If it was the warning being missed, where should it be relocated to to help others? (So I can improve documents)

>Should I flash .hex file or .bin file?

Use the `bootloader.hex` file. Runtime is just for when you want to update the bootloader, and for backing up the original one. If you use SWD you can just straight to it.

**Comment by AreYouLoco** [comment]
Hi @Ralim first of all thank you for IronOS. I am happy user for many years and I do follow releases. Well TL;DR I was doing bootloader update at 2am drunk😅 without paying to much attention what I was doing as I just believed what I remember from when I switched to your version of DFU instead of re-read the warnings and documentation. I tried to flash bootloader from running bootloader few times. Without runtime stage. Its still detected with dfu-tools. But there is no way to flash anything now so I will just use ST-Link to recover. Documentation is fine when you read it before actually doing stuff. Maybe photos posted here with pinout for recovery could be added to documentation. With the mention of SWD pind being or not connected to D- D+. And some table with hardware revisions to be filled up if the users report on which rev is what.

> did you flash the board without heeding the warning that on newer units dont let you do this

It is not new unit. I bought it long time ago. As I mentioned rev 2.47A. So I am certain it was without that stupid protection of newer units in place. I should have backup somewhere on the drive archive but anyway I checked checksum and it was known version. It was already running some older release of your bootloader I just tried to update it yesterday because why not.

So just to confirm when I what to flash with st-link tools I use .hex file? What is the .bin for then?

**Comment by Ralim** [comment]
> Well TL;DR I was doing bootloader update at 2am drunk

We have all been there doing something at a not-the-wisest-time.

>I tried to flash bootloader from running bootloader few times.

Aaaah, yeah that _should_ be blocked now, I added that in later revs of this bootloader.


>Maybe photos posted here with pinout for recovery could be added to documentation. With the mention of SWD pind being or not connected to D- D+. And some table with hardware revisions to be filled up if the users report on which rev is what.

Good idea


>So just to confirm when I what to flash with st-link tools I use .hex file? What is the .bin for then?

Basically the idea is use `.dfu` if supported; if not use `.hex`; and lastly `.bin`.
Problem is that `.bin` files dont tell you what address to flash at, so then you need to figure that out. I just provide all the options just-in-case.

**Comment by AreYouLoco** [comment]
I think later on today I will atempt recovery. I will use linux st-link tools and last time I used it I had to provide address to which the binary should be written.

So dfu when using dfu-tool. Hex when putting a file in the mass storage (or also when flashing with st-link tools?) And binary as last resort also st-link tools or some other external programmer. Right? I am sorry but I am still confused if I should flash hex or bin.

Anyway 50% chance will try both. Its not like it will brick even more just wont work(boot).

**Comment by Ralim** [comment]
Yeah your right, `.dfu` if using `dfu-util` or other software that supports it.
Otherwise hex is the goto. `bin` is mostly just for last resort. Using the `.hex` means you wont need to specify any address information

**Comment by AreYouLoco** [comment]
@kenblu24 thx for the pinout. Managed to unbrick mine.
